src/main/java/org/tron/common/runtime/vm/program/InternalTransaction.java

```diff
@@ -115,87 +137,67 @@ public String getNote() {
-//        this.sendAddress = transaction.get(2).getRLPData();
-//        setReceiveAddress(transaction.get(3).getRLPData());
-//        setValue(transaction.get(4).getRLPData());
-//        setGasPrice(transaction.get(5).getRLPData());
-//        setGasLimit(transaction.get(6).getRLPData());
-//        setData(transaction.get(7).getRLPData());
-//        this.note = new String(transaction.get(8).getRLPData());
-//        this.deep = decodeInt(transaction.get(9).getRLPData());
-//        this.index = decodeInt(transaction.get(10).getRLPData());
-//        this.rejected = decodeInt(transaction.get(11).getRLPData()) == 1;
-//        */
-//        this.parsed = true;
-//    }
-
-
-  private static byte[] intToBytes(int value) {
-    return ByteBuffer.allocate(Integer.SIZE / Byte.SIZE)
-        .order(ByteOrder.LITTLE_ENDIAN)
-        .putInt(value)
-        .array();
-  }
-
-  private static int bytesToInt(byte[] bytes) {
-    return isEmpty(bytes) ? 0 : ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN).getInt();
-  }
-
-    /*
-    private static byte[] encodeInt(int value) {
-        return RLP.encodeElement(intToBytes(value));
-    }
-    */
+  public byte[] getData() {
+    protoParse();
+    return data;
+  }
 
-  private static int decodeInt(byte[] encoded) {
-    return bytesToInt(encoded);
+  protected void setValue(long value) {
+    this.value = value;
+    parsed = true;
   }
 
-  @Override
-  public ECKey getKey() {
-    throw new UnsupportedOperationException("Cannot sign internal transaction.");
+  public byte[] getReceiveAddress() {
+    protoParse();
+    return receiveAddress;
   }
 
-  @Override
-  public void sign(byte[] privKeyBytes) throws ECKey.MissingPrivateKeyException {
-    throw new UnsupportedOperationException("Cannot sign internal transaction.");
+  private void protoParse() {
+    if (parsed) {
+      return;
+    }
+    try {
+      this.hash = Hash.sha3(protoEncoded);
+      this.parsed = true;
+    } catch (Exception e) {
+      throw new RuntimeException("Error on parsing proto", e);
+    }
+  }
+
+  public byte[] getHash() {
+    if (!isEmpty(hash)) {
+      return Arrays.copyOf(hash, hash.length);
+    }
+
+    protoParse();
+    byte[] plainMsg = this.getEncoded();
+    return Hash.sha3(plainMsg);
+  }
+
+
+  public byte[] getEncoded() {
+
+    if (protoEncoded != null) {
+      if (null == this.hash) {
+        this.hash = Hash.sha3(protoEncoded);
+      }
+      return protoEncoded;
+    }
+
+    byte[] valueByte = Longs.toByteArray(this.value);
+    byte[] raw = new byte[this.receiveAddress.length + this.data.length + valueByte.length];
+    System.arraycopy(this.receiveAddress, 0, raw, 0, this.receiveAddress.length);
+    System.arraycopy(this.data, 0, raw, this.receiveAddress.length, this.data.length);
+    System.arraycopy(valueByte, 0, raw, this.data.length, valueByte.length);
+    this.protoEncoded = raw;
+    this.hash = Hash.sha3(protoEncoded);
+
+    return protoEncoded;
```

Comment:
![Vulnerability](https://sonarsource.github.io/sonar-github/vulnerability.png 'Vulnerability') Vulnerability: Return a copy of "protoEncoded". ([squid:S2384](https://sonarcloud.io/organizations/tron-zhaohong/rules#rule_key=squid%3AS2384))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=java-tron&pullRequest=1220&issues=AWT01HPBvlEfaeH0aL_g&open=AWT01HPBvlEfaeH0aL_g)

[ISSUE_KEY]: AWT01HPBvlEfaeH0aL_g